### PR TITLE
Experiment with partializing functions

### PR DESCRIPTION
**Pull request by JLockerman**
With this commit running

    select partialize(avg(i)) from foo;

works, as does for every other partializable aggregate postgres provides, and returns the bytea serialized form of the aggregate context. Attempting to mix partialized and non-partialized aggregates does _not_ work, but I'm not sure we care at the moment.

### CODE REVIEW

**Comment by JLockerman** [comment]
Example explain output for `explain (verbose) select partialize(avg(i)) from _timescaledb_internal._hyper_1_2_chunk;`
```
                                           QUERY PLAN                                           
------------------------------------------------------------------------------------------------
 Partial Aggregate  (cost=41.88..41.89 rows=1 width=32)
   Output: partialize(PARTIAL avg(i))
   ->  Seq Scan on _timescaledb_internal._hyper_1_2_chunk  (cost=0.00..35.50 rows=2550 width=4)
         Output: i
(4 rows)
```

File: src/plan_partialize.c [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+TS_FUNCTION_INFO_V1(ts_partialize);
+
+/*
+ * the partialize function mainly serves as a marker that the aggregate called
+ * within should return a partial instead of a result. Most of the actual work
+ * occurs in the planner, with the actual function just used to ensure the
+ * return type is correct.
+ */
+TSDLLEXPORT Datum
+ts_partialize(PG_FUNCTION_ARGS)
```
Comment (davidkohn88) [inline src/plan_partialize.c:32]: So, I'm thinking that we're going to need a departialize function or something similar as well.

**Review by davidkohn88 — COMMENTED** [review]
good start.

File: src/plan_partialize.c (hunk 1 above)
Comment (JLockerman) [reply src/plan_partialize.c:32]: `departialize` is much more difficult as it requires knowing ahead of time what the internal state we're deserializing to is. It'll likely be much more invasive.

**Comment by JLockerman** [comment]
**NOTE** I'm not entirely certain this is correct with `min` and `max` as postgres optimizes those specially

**Comment by JLockerman** [comment]
Just pushed a deserialze function which works like so
```
postgres=# select partialize(avg(i)) from _timescaledb_internal._hyper_1_2_chunk;
                                         partialize                                         
--------------------------------------------------------------------------------------------
 \x0000000100000000000000140000000200000001000000080000000000000032000000080000000000000e8d
(1 row)

postgres=# select aggregate_transition_type('avg(int)');
 aggregate_transition_type 
---------------------------
 bigint[]
(1 row)

postgres=# select aggregate_deserialize('\x0000000100000000000000140000000200000001000000080000000000000032000000080000000000000e8d', 'avg(int)', NULL::bigint[]);
WARNING:  cache reference leak: cache pg_aggregate (0), tuple 0/2 has count 1
 aggregate_deserialize 
-----------------------
 {50,3725}
(1 row)
```
   
It calls the aggregate's deserialize function if it exists, otherwise it calls the binary input function for the transition type.  
**TODO** I'm not sure yet if this is the most useful behavior, it may pay to only deserialize if it's using a builtin transition type (ie. one which would _not_ normally be turned into a `bytea`) and let postgres handle everything from there.

**Comment by JLockerman** [comment]
also added functions
```SQL
aggregate_transition_type(REGPROCEDURE)
```
which returns the transition type of an aggregate and
```SQL
aggregate_deserialize_fn(REGPROCEDURE)
```
which returns the deserialization function the aggregate uses, if any.

File: src/plan_partialize.c [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+		elog(ERROR, "invalid return type, got %d expected %d", arg_type, aggform->aggtranstype);
+
+	if (OidIsValid(aggform->aggdeserialfn))
+	{
+		FmgrInfo	deserialize_finfo;
+
+		fmgr_info(aggform->aggdeserialfn, &deserialize_finfo);
+
+		deserialized = FunctionCall1Coll(&deserialize_finfo, InvalidOid, PointerGetDatum(serialized));
+	}
```
Comment (JLockerman) [inline src/plan_partialize.c:220]: If we decide we want postgres to handle deserialization for `internal` datatypes, we simply remove the `if` part of this conditional.
(I suspect this _will_ ultimately end up being the strategy, but we need to dig a little deeper into `Finalize Aggregate` to be sure)

**Pull request closed without merging**